### PR TITLE
Add using for extension get enumerator

### PR DESCRIPTION
**Pull request by YairHalberstadt**
extends the add using code fix to import an extension GetEnumerator or GetAsyncEnumerator

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 1]
```diff
@@ -656,6 +656,77 @@ private static bool VerifyGetAwaiter(IMethodSymbol getAwaiter)
             return methods.Any(m => m.Name == WellKnownMemberNames.GetResult && !m.Parameters.Any());
         }
 
+        public static bool IsValidGetEnumerator(this IMethodSymbol symbol)
+            => symbol.Name == WellKnownMemberNames.GetEnumeratorMethodName &&
+            VerifyGetEnumerator(symbol);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:661]: Align Verify with symbol

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 2]
```diff
@@ -656,6 +656,77 @@ private static bool VerifyGetAwaiter(IMethodSymbol getAwaiter)
...
+                return false;
+            }
+
+            // bool MoveNext()
+            if (!members.OfType<IMethodSymbol>().Any(x => x is
+            {
+                Name: WellKnownMemberNames.MoveNextMethodName,
+                ReturnType: { SpecialType: SpecialType.System_Boolean },
+                Parameters: { Length: 0 },
+            }))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:688]: Indent this. Or extract local

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 3]
```diff
@@ -656,6 +656,77 @@ private static bool VerifyGetAwaiter(IMethodSymbol getAwaiter)
...
+            // T Current { get }
+            if (!members.OfType<IPropertySymbol>().Any(p => p.Name == WellKnownMemberNames.CurrentPropertyName && p.GetMethod != null))
+            {
+                return false;
+            }
+
+            // Task<bool> MoveNext()
+            // We don't check for the return type, since it can be any awaitable wrapping a boolean, 
+            // which is too complex to be worth checking here
+            if (!members.OfType<IMethodSymbol>().Any(x => x.Name == WellKnownMemberNames.MoveNextAsyncMethodName))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:722]: Why not check parameters like you do above?

Comment (YairHalberstadt) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:722]: MoveNextAsync allows optional parameters

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:722]: doc plz

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 2 above)
Comment (YairHalberstadt) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:688]: I get a formatting error when I indent it. Extracting a local makes it ugly enough I just wouldn't use patterns at all.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 1 above)
Comment (YairHalberstadt) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:661]: That's inconsistent with the style above in the document:

https://github.com/dotnet/roslyn/blob/b0787c0654d8fa40a9a5ed6a1a218106c07de72b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs#L623-L625

Should I change them both, or leave it.

File: src/Features/Core/Portable/AddImport/SymbolReferenceFinder.cs [hunk 4]
```diff
@@ -484,6 +486,50 @@ private async Task<ImmutableArray<SymbolReference>> GetReferencesForGetAwaiterAs
...
+            /// <see cref="SymbolReference"/>s to the <see cref="INamespaceSymbol"/>s that contain
+            /// the static classes that those extension methods are contained in.
+            /// </summary>
+            private async Task<ImmutableArray<SymbolReference>> GetReferencesForGetEnumeratorAsync(SearchScope searchScope)
+            {
+                searchScope.CancellationToken.ThrowIfCancellationRequested();
+
+                if (_owner.CanAddImportForGetEnumerator(_diagnosticId, _syntaxFacts, _node))
+                {
+                    var type = _owner.GetCollectionExpressionInfo(_semanticModel, _syntaxFacts, _node);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/SymbolReferenceFinder.cs:500]: GetCollectionExpressionType ?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 5]
```diff
@@ -656,6 +656,82 @@ private static bool VerifyGetAwaiter(IMethodSymbol getAwaiter)
...
+
+        private static bool VerifyGetEnumerator(IMethodSymbol getEnumerator)
+        {
+            var returnType = getEnumerator.ReturnType;
+            if (returnType == null)
+            {
+                return false;
+            }
+
+            var members = returnType.AllInterfaces.Concat(returnType.GetBaseTypesAndThis())
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:671]: what if it doesn't implement an interface?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 6]
```diff
@@ -656,6 +656,82 @@ private static bool VerifyGetAwaiter(IMethodSymbol getAwaiter)
...
+            // bool MoveNext()
+            if (!members.OfType<IMethodSymbol>().Any(x =>
+            {
+                var isMoveNext = x is
+                {
+                    Name: WellKnownMemberNames.MoveNextMethodName,
+                    ReturnType: { SpecialType: SpecialType.System_Boolean },
+                    Parameters: { Length: 0 },
+                };
+                return isMoveNext;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:691]:
```suggestion
                return x is
                {
                    Name: WellKnownMemberNames.MoveNextMethodName,
                    ReturnType: { SpecialType: SpecialType.System_Boolean },
                    Parameters: { Length: 0 },
                };
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 7]
```diff
@@ -656,6 +656,82 @@ private static bool VerifyGetAwaiter(IMethodSymbol getAwaiter)
...
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public static bool IsValidGetAsyncEnumerator(this IMethodSymbol symbol)
+            => symbol.Name == WellKnownMemberNames.GetAsyncEnumeratorMethodName &&
+            VerifyGetAsyncEnumerator(symbol);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:702]:
```suggestion
                VerifyGetAsyncEnumerator(symbol);
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 5 above)
Comment (YairHalberstadt) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:671]: I have a test for that case.

**Comment by YairHalberstadt** [comment]
@CyrusNajmabadi Could you have another look at this please?
And could I have a second review so we can get this merged.

File: src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs [hunk 8]
```diff
@@ -544,10 +546,20 @@ private static ITypeSymbol GetAwaitInfo(SemanticModel semanticModel, ISyntaxFact
             return semanticModel.GetTypeInfo(innerExpression).Type;
         }
 
+        private static ITypeSymbol GetCollectionExpressionType(SemanticModel semanticModel, ISyntaxFacts syntaxFactsService, SyntaxNode node)
+        {
+            var collectionExpression = FirstForeachCollectionExpressionAncester(syntaxFactsService, node);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs:551]:
```suggestion
            var collectionExpression = FirstForeachCollectionExpressionAncestor(syntaxFactsService, node);
```

**Comment by CyrusNajmabadi** [comment]
Will merge once we get a successful test pass :)

**Comment by YairHalberstadt** [comment]
I think we may have to merge master into the target branch to get tests to pass

**Comment by CyrusNajmabadi** [comment]
Which compiler dev is working on this?  I can ask them to do that.

**Comment by YairHalberstadt** [comment]
@333fred

**Comment by 333fred** [comment]
Added to my list for the morning.

**Comment by 333fred** [comment]
Submitted https://github.com/dotnet/roslyn/pull/46212 to merge master into this branch.

**Comment by YairHalberstadt** [comment]
Closing and reopening to trigger CI

**Comment by YairHalberstadt** [comment]
@CyrusNajmabadi Everything is passing except for the test UpdaterService in Windows Desktop Release 32 bit.

It looks like this is commonly failing on PRs targeting master, and not specific to this PR.

**Comment by 333fred** [comment]
@CyrusNajmabadi is this ready for merge?

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**